core/src/impl/Kokkos_Atomic_Compare_Exchange_Strong.hpp

```diff
@@ -381,7 +381,7 @@ KOKKOS_INTERNAL_INLINE_DEVICE_IF_CUDA_ARCH bool _atomic_compare_exchange_strong(
     T* dest, T compare, T val, MemoryOrderSuccess, MemoryOrderFailure,
     typename std::enable_if<
         (sizeof(T) == 1 || sizeof(T) == 2 || sizeof(T) == 4 ||
-         sizeof(T) == 8) &&
+         sizeof(T) == 8 || sizeof(T) == 16) &&
```

Comment:
why this change?

Reply:
Wow seem it doesn't. This was my attempt of a clean fix for #3344. 
This is the msvc (clang-cl) error:
```
2>C:\Code\kokkos\core\src\impl/Kokkos_Atomic_Generic.hpp(210,16): error : no matching function for call to 'atomic_compare_exchange'
2>C:\Code\kokkos\core\src\impl/Kokkos_Atomic_Generic.hpp(556,16): message : in instantiation of function template specialization 'Kokkos::Impl::atomic_fetch_oper<Kokkos::Impl::AddOper<long long, const long long>, long long>' requested here
2>C:\Code\kokkos\core\src\impl/Kokkos_Atomic_Increment.hpp(132,11): message : in instantiation of function template specialization 'Kokkos::atomic_fetch_add<long long>' requested here
2>C:\Code\kokkos\core\src/impl/Kokkos_Atomic_Compare_Exchange_Strong.hpp(154,12): message : candidate function not viable: no known conversion from 'unsigned long long *' to 'volatile int *const' for 1st argument
2>C:\Code\kokkos\core\src/impl/Kokkos_Atomic_Compare_Exchange_Strong.hpp(162,13): message : candidate function not viable: no known conversion from 'unsigned long long *' to 'volatile long *const' for 1st argument
2>C:\Code\kokkos\core\src/impl/Kokkos_Atomic_Compare_Exchange_Strong.hpp(174,21): message : candidate function not viable: no known conversion from 'unsigned long long *' to 'volatile unsigned int *const' for 1st argument
2>C:\Code\kokkos\core\src/impl/Kokkos_Atomic_Compare_Exchange_Strong.hpp(180,22): message : candidate function not viable: no known conversion from 'unsigned long long *' to 'volatile unsigned long *const' for 1st argument
2>C:\Code\kokkos\core\src/impl/Kokkos_Atomic_Compare_Exchange_Strong.hpp(189,10): message : candidate template ignored: requirement 'sizeof(unsigned long long) == sizeof(int)' was not satisfied [with T = unsigned long long]
2>C:\Code\kokkos\core\src/impl/Kokkos_Atomic_Compare_Exchange_Strong.hpp(208,10): message : candidate template ignored: requirement 'sizeof(unsigned long long) != sizeof(int) && sizeof(unsigned long long) == sizeof(long)' was not satisfied [with T = unsigned long long]
2>C:\Code\kokkos\core\src/impl/Kokkos_Atomic_Compare_Exchange_Strong.hpp(253,10): message : candidate template ignored: requirement '(sizeof(unsigned long long) != 4) && (sizeof(unsigned long long) != 8)' was not satisfied [with T = unsigned long long]
```

Reply:
If you're okay with this - I'll implement a define to distinguish between `clang++` and `clang-cl,` because `msvc` gets activated when `_MSC_VER` is found. 
```
#if defined(_MSC_VER) && !defined(KOKKOS_COMPILER_INTEL)
#define KOKKOS_COMPILER_MSVC _MSC_VER
#endif
```

Reply:
Tried to fix this using c9535e4, see below.